### ISSUE TITLE
Treatment of discrepancies in Mach-O slices of fat binaries

### ISSUE THREAD

**Issue by UniqMartin** [Homebrew/ruby-macho#50]
As has been raised in https://github.com/Homebrew/brew/pull/592#issuecomment-235796041 and subsequent comments, test bots running `ruby-macho` fail to bottle `alot` from PR Homebrew/homebrew-core#1663:

```console
$ brew bottle --verbose --debug --json alot
/opt/brewery/dummy/Library/Homebrew/brew.rb (Formulary::TapLoader): loading /opt/brewery/dummy/Library/Taps/homebrew/homebrew-core/Formula/alot.rb
==> Determining alot bottle revision...
==> Bottling alot-0.3.7.el_capitan.bottle.tar.gz...
Changing install name in /opt/brewery/dummy/Cellar/alot/0.3.7/libexec/lib/python2.7/site-packages/gpgme/_gpgme.so
  from /opt/brewery/dummy/opt/gpgme/lib/libgpgme.11.dylib
    to @@HOMEBREW_PREFIX@@/opt/gpgme/lib/libgpgme.11.dylib
Error: No such dylib name: /opt/brewery/dummy/opt/gpgme/lib/libgpgme.11.dylib
/opt/brewery/dummy/Library/Homebrew/vendor/macho/macho/macho_file.rb:240:in `change_install_name'
/opt/brewery/dummy/Library/Homebrew/vendor/macho/macho/fat_file.rb:171:in `block in change_install_name'
/opt/brewery/dummy/Library/Homebrew/vendor/macho/macho/fat_file.rb:170:in `each'
/opt/brewery/dummy/Library/Homebrew/vendor/macho/macho/fat_file.rb:170:in `change_install_name'
/opt/brewery/dummy/Library/Homebrew/vendor/macho/macho/tools.rb:33:in `change_install_name'
/opt/brewery/dummy/Library/Homebrew/os/mac/ruby_keg.rb:13:in `change_install_name'
[…snip…]
```

The reason turns out to be that the problematic file is a fat binary, but not all of its slices link to the mentioned dylib (most likely because the dylib in question only has a single architecture):

```console
$ otool -arch all -L /opt/brewery/dummy/Cellar/alot/0.3.7/libexec/lib/python2.7/site-packages/gpgme/_gpgme.so
/opt/brewery/dummy/Cellar/alot/0.3.7/libexec/lib/python2.7/site-packages/gpgme/_gpgme.so (architecture i386):
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1226.10.1)
/opt/brewery/dummy/Cellar/alot/0.3.7/libexec/lib/python2.7/site-packages/gpgme/_gpgme.so (architecture x86_64):
	/opt/brewery/dummy/opt/gpgme/lib/libgpgme.11.dylib (compatibility version 26.0.0, current version 26.0.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1226.10.1)

$ otool -arch all -L /opt/brewery/dummy/opt/gpgme/lib/libgpgme.11.dylib
/opt/brewery/dummy/opt/gpgme/lib/libgpgme.11.dylib:
	/opt/brewery/dummy/opt/gpgme/lib/libgpgme.11.dylib (compatibility version 26.0.0, current version 26.0.0)
	/opt/brewery/dummy/opt/libassuan/lib/libassuan.0.dylib (compatibility version 8.0.0, current version 8.3.0)
	/opt/brewery/dummy/opt/libgpg-error/lib/libgpg-error.0.dylib (compatibility version 20.0.0, current version 20.1.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1226.10.1)
```

@tdsmith This didn't happen before the formula in the PR was switched over to the `virtualenv` system. It strikes me as broken that this is creating extensions that pretend to be universal, but the 32-bit slice is probably dysfunctional. Can you take a closer look at the Python/`virtualenv` side of things and check what the correct behavior should be?

My conclusion for `ruby-macho` is that we need to handle these situations more resiliently. It's unusual, but perfectly legal, for different Mach-O slices of a fat binary to have a different set of dylibs they link to. `install_name_tool` seems to handle this correctly, thus we should do that, too. (Though we should continue being a bit more strict within reasonable bounds.)

**TL;DR/practical suggestion:** By default, consider a modification to a fat binary a success when it succeeds for one of its Mach-O slices. Optionally offer a strict mode that would require the change to succeed for all slices (current behavior), e.g. via:

```ruby
file.change_install_name(old_path, new_path, :strict => true)
```

cc @woodruffw

**Comment by woodruffw** [comment]
Yep, I was worried that this would show up eventually. In general, all of the `FatFile` LC modification methods assume that their constituent `MachOFile`s are identical - a bad assumption considering that nothing in the Mach-O spec requires fat-file constituents to even be similar to each other.

I think having a `:strict` option (`false` by default) would cover our bases nicely, good idea!

I'll open a PR for this in a moment :+1:

**Comment by woodruffw** [comment]
Hmm, I think I'll just add the options hash to both `MachOFile` and `FatFile` to preserve uniformity (and prevent additional complexity in `MachO::Tools`) - it just won't have any effect on single-arch.

**Comment by UniqMartin** [comment]
> Hmm, I think I'll just add the options hash to both `MachOFile` and `FatFile` to preserve uniformity (and prevent additional complexity in `MachO::Tools`).

Yes, I'm definitely also in favor of handling it like that.
